api_handlers.go

```diff
@@ -120,6 +132,56 @@ func apiTestRunHandler(w http.ResponseWriter, r *http.Request) {
 	w.Write(testRunsBytes)
 }
 
+// apiTestRunPostHandler is responsible for handling TestRun submissions (via HTTP POST requests).
+// It asserts the presence of a required secret token, then saves the JSON blob to the Datastore.
+// See models.go for the JSON format expected.
+func apiTestRunPostHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	var err error
+
+	// Fetch pre-uploaded Token entity.
+	suppliedSecret := r.URL.Query().Get("secret")
+	tokenKey := datastore.NewKey(ctx, "Token", "upload-token", 0, nil)
+	var token Token
+	if err = datastore.Get(ctx, tokenKey, &token); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if suppliedSecret != token.Secret {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	var body []byte
+	if body, err = ioutil.ReadAll(r.Body); err != nil {
```

Comment:
I think you need to add `defer r.Body.Close()` somewhere in this function in order to not leak sockets.

Reply:
Done

Reply:
Ah! TIL :+1: